### PR TITLE
Added resize flag handlers for imread and imdecode | Issue #16203

### LINKED ISSUE #16203: imdecode ignores IMREAD_REDUCED_* flags
Passing any of the `IMREAD_REDUCED_` flags, e.g. `IMREAD_REDUCED_GRAYSCALE_2`, to the flags parameter of `imdecode()` has no effect. I notice the `imread_()` function includes this code:

    int scale_denom = 1;
    if( flags > IMREAD_LOAD_GDAL )
    {
    if( flags & IMREAD_REDUCED_GRAYSCALE_2 )
        scale_denom = 2;
    else if( flags & IMREAD_REDUCED_GRAYSCALE_4 )
        scale_denom = 4;
    else if( flags & IMREAD_REDUCED_GRAYSCALE_8 )
        scale_denom = 8;
    }

    /// set the scale_denom in the driver
    decoder->setScale( scale_denom );

This code is absent in `imdecode()` (and `imdecode_`).

For those that aren't familiar with that flag: The expected effect is that the image is resized by the stated factor if it is a JPEG. This happens for "free" due to the way the JPEG encoding works, so it is much faster than opening the image at full resolution and then doing a separate resize.

### PR DESCRIPTION
**Pull request by ganesh-k13**
# This pullrequest changes

1. Added flag handler for imdecode which was the issue
2. Take care of color resize cases
3. Port changes to imread as well.

resolves #16203

```
force_builders=Win32
```

### CODE REVIEW

**Comment by ganesh-k13** [comment]
Hi @sturkmen72, I have added the handlers as discussed. In addition to this I noticed that color resize was not taking place in case of imread, I fixed that as well.

**Comment by sturkmen72** [comment]
i believe your addition is not correct ( or unnecessary)

**Comment by sturkmen72** [comment]
also see https://github.com/opencv/opencv/blob/master/modules/imgcodecs/src/loadsave.cpp#L494 should be added in imdecode

**Comment by ganesh-k13** [comment]
Hi @sturkmen72 thanks for the response, so the addition to imread is not needed? Or even imdecode is not necessary? As mentioned in last comment, do I add that to imdecode?

File: modules/imgcodecs/src/loadsave.cpp [hunk 1]
```diff
@@ -421,14 +421,11 @@ imread_( const String& filename, int flags, int hdrtype, Mat* mat=0 )
...
     {
-    if( flags & IMREAD_REDUCED_GRAYSCALE_2 )
-        scale_denom = 2;
-    else if( flags & IMREAD_REDUCED_GRAYSCALE_4 )
-        scale_denom = 4;
-    else if( flags & IMREAD_REDUCED_GRAYSCALE_8 )
-        scale_denom = 8;
+        /// Make it an even number as color flags are odd
+        /// We scale the image based on resize factor
+        scale_denom = ( flags % 2 == 0 ) ? flags >> 3 : ( flags - 1 ) >> 3;
```
Comment (alalek) [inline modules/imgcodecs/src/loadsave.cpp]: Not readable.
Not maintainable (after enum refactoring code may become broken).

Comment (ganesh-k13) [reply modules/imgcodecs/src/loadsave.cpp]: Thanks for the review @alalek, do I add the if conditions for greyscale and color, then it'll be readable, also I add these conditions to imread as well right?

File: modules/imgcodecs/src/loadsave.cpp [hunk 2]
```diff
@@ -421,14 +421,14 @@ imread_( const String& filename, int flags, int hdrtype, Mat* mat=0 )
     }
 
     int scale_denom = 1;
-    if( flags > IMREAD_LOAD_GDAL )
+    if( ( flags > IMREAD_LOAD_GDAL ) & ( flags < IMREAD_IGNORE_ORIENTATION ) )
```
Comment (sturkmen72) [inline modules/imgcodecs/src/loadsave.cpp]: if( flags > IMREAD_LOAD_GDAL ) is good here ( to make usage like `IMREAD_REDUCED_GRAYSCALE_4 | IMREAD_IGNORE_ORIENTATION`  acceptable )
I recommend that you do not make any changes to the `imread ` function

Comment (ganesh-k13) [reply modules/imgcodecs/src/loadsave.cpp]: Sure @sturkmen72, I'll remove it.

File: modules/imgcodecs/src/loadsave.cpp [hunk 3]
```diff
@@ -766,6 +766,20 @@ imdecode_( const Mat& buf, int flags, int hdrtype, Mat* mat=0 )
     if( !decoder )
         return 0;
 
+    int scale_denom = 1;
+    if( ( flags > IMREAD_LOAD_GDAL ) & ( flags < IMREAD_IGNORE_ORIENTATION ) )
```
Comment (sturkmen72) [inline modules/imgcodecs/src/loadsave.cpp]: if( flags > IMREAD_LOAD_GDAL )

Comment (ganesh-k13) [reply modules/imgcodecs/src/loadsave.cpp]: Oh sorry, my bad, skipped this, thanks.

File: modules/imgcodecs/test/test_read_write.cpp [hunk 4]
```diff
@@ -5,6 +5,151 @@
 
 namespace opencv_test { namespace {
 
+TEST(Imgcodecs_Image, imread_reduce_flags)
```
Comment (alalek) [inline modules/imgcodecs/test/test_read_write.cpp]: Please add `#ifdef HAVE_JPEG` guard for these tests (to avoid test failures in some configurations).

Comment (ganesh-k13) [reply modules/imgcodecs/test/test_read_write.cpp]: Sure @alalek, will do

**Comment by sturkmen72** [comment]
`imread `manually resizes other formats than jpeg see https://github.com/opencv/opencv/blob/master/modules/imgcodecs/src/loadsave.cpp#L494-L497

**Comment by ganesh-k13** [comment]
@sturkmen72, thanks for that, shall I add that to imdecode as well as you showed earlier?

**Comment by alalek** [comment]
@ganesh-k13 Please fix warnings from Win32 builder (add `static_cast<size_t>()`)

**Comment by sturkmen72** [comment]
@alalek let me do some tests before merge

**Comment by ganesh-k13** [comment]
> `imread `manually resizes other formats than jpeg see https://github.com/opencv/opencv/blob/master/modules/imgcodecs/src/loadsave.cpp#L494-L497

@sturkmen72, do we make this in a different PR, or shall I add it to `imdecode` now?

**Comment by ganesh-k13** [comment]
Hey @sturkmen72 / @alalek, I'm getting cores if we add that if conditions and try other formats, if that if is not there then resize does not happen but no crash. It involves some work, can I raise a different PR for that, might involve more work. Please advise.

**Comment by sturkmen72** [comment]
you can add 
```
    if( decoder->setScale( scale_denom ) > 1 ) // if decoder is JpegDecoder then decoder->setScale always returns 1
    {
        resize( mat, mat, Size( size.width / scale_denom, size.height / scale_denom ), 0, 0, INTER_LINEAR_EXACT);
    }
   decoder.release();
``` 
end of imdecode_  
( delete `decoder.release();` which located a few lines upper )

**Comment by ganesh-k13** [comment]
Ohh thanks, I didn't see it was getting freed. I'll add that and test.

**Comment by ganesh-k13** [comment]
@sturkmen72 thanks for the fix, it works. @alalek I have added a case for the same by resizing a PNG to hit that code path.

**Comment by ganesh-k13** [comment]
@sturkmen72 / @alalek, any updates on this one? any changes needed?

**Comment by ganesh-k13** [comment]
Thanks for the link @asmorkalov, I'll make those changes

**Comment by alalek** [comment]
Parametrized tests look like overkill here. Duplicated code and checks may be moved into common function, like [this](https://github.com/opencv/opencv/blob/4.2.0/modules/core/test/test_eigen.cpp#L390-L455) (important note, that return value should be 'void').

**Comment by ganesh-k13** [comment]
Hi @alalek , @sturkmen72 , @asmorkalov, we can see that imread and imdecode have duplicate code in both test and src. I would suggest we make it a different PR as we have digressed from the initial issue. I can raise a new issue and PR for that.

**Comment by ganesh-k13** [comment]
Hi @alalek , @asmorkalov , I went ahead with refactor of tests anyways. I went with `Value-Parameterized Tests` as I wanted to try something new and it seems very well suited for our case. Please merge this PR if changes are satisfactory.

**Comment by asmorkalov** [comment]
@ganesh-k13 Well done! Please squash your commits to have clean history and we can merge the patch. @alalek What do you think?

**Pull request merged**